libs/@local/hql/README.md

```diff
@@ -0,0 +1,51 @@
+# HQL - HASH Query Language
+
+This is the compiler, planner and executor for the HASH Query Language (HQL). HQL is a functional programming language that is created using a subset of JSON.
+
+## Language
```

Comment:
It's super hard to follow what the CST is representing without a grammar being specified but I'm happy to add this in a follow up. This is also the point where we should mostly agree on a grammar.

On the grammar itself I have a few questions:
- Why we'll have special forms in the future at all?
- Why the CST as presented a few lines beblow does not fully align with the grammar as specified in Notion? I would expect the CST to fully align with the EBNF, e.g. a signature would be
	```ts
	{
		type: "signature",
		generics: {
			ident: Ident,
			where: Type[]
		}[],
		arguments: {
			ident: Ident,
			type: Type
		}[],
		return?: Type
	}
	```

Reply:
I can also add the syntax to the README itself, right now it is specified above the parse function for the different constructs. I had to change it a little from the one in the Notion document, to resolve potential parse ambiguities.

Reply:
> Why we'll have special forms in the future at all?

Which special form? Special form values or special form functions?

Special form functions are needed to special case certain constructs, functional languages like e.g. closure have them as well.

So to say, we need to special case in the AST or during analysis certain constructs to make our lives easier, or to be able to delay execution.

For example:

```
["do", ...exprs]
```

is a special form function, that denotes a new block of expressions, that introduce a new scoped block (for variables and such) and is lazily computed.

Same thing e.g. `ifElse` and such, where an expression is only evaluated it the condition is either true of false.

Special form values on the other hand are just an idea to make authoring CST easier, the bridge the CST has to "wander" is a perilous one, because we need to consolidate two different use-cases, machine authored queries and human authored queries.
Because we foresee that in the short/mid-term this CST representation is going to be used as the interface, instead of something like GQL or SPARQL, we need to make it possible to easily author that content, while still having a way for us to later, easily implement a e.g. GQL transpiler.

This motivates many of the choices done here, where you have an implicit representation, that is easier to author, and an explicit one, that is easier for tools.

e.g. it's a lot more concise (also in indentation to write):

```json
["do",
	["let", "a", {"const": 2}],
	["let", "b", {"const" 3}],
	["let", "c", ["add", "a", "b"]]
	"c"
]
```

than the "full machine representation":

```json
{
  "fn": "do",
  "args": [
    {
      "fn": { "var": "let" },
      "args": [{ "var": "a" }, { "const": 2, "type": "Integer" }]
    },
    {
      "fn": { "var": "let" },
      "args": [{ "var": "b" }, { "const": 3, "type": "Integer" }]
    },
    {
      "fn": { "var": "let" },
      "args": [
        { "var": "c" },
        { "fn": "add", "args": [{ "var": "a" }, { "var": "b" }] }
      ]
    },
    { "var": "c" }
  ]
}
```

while the latter is a lot easier to generate for a computer program, it is pain to author as a human, something we expect to happen.

Reply:
special form values go into the same direction, by enabling us to be even more concise:

```json
["do",
	["let", "a", "#value(2)"],
	["let", "b", "#value(3)"],
	["let", "c", ["add", "a", "b"]]
	"c"
]
```

Reply:
This also motivates the difference between the object representation, and the more implicit terse representation:

```json
{"sig": "() -> Unit"} // or the full decoded version
```

is easier to author without the fear of ambiguities, than:

```json
"() -> Unit"
```